### DISCUSSION TITLE
Priorities for new language features

### DISCUSSION THREAD

**Discussion by titzer** [titzer/virgil#122] [Ideas]
This is a thread to discuss some ideas for how to improve Virgil from a language design and productivity perspective. I'd like to hear from people who have used the language and what they might find interesting in coming versions. Some features are easier to implement than others, and given that I'm spending most of my time on Wizard and only working on Virgil in the capacity to make it a better *systems language*, some of these might be more aspirational than others without help.

## Generators

One goal is to be able write libraries that are easy to use with terse code. Iteration over something like a `Vector` should look like iterating over an array or a list. E.g.

```
var x: Array<byte>;
for (b in x) { ... } // iterating over array is supported in "for in"
var v: Vector<byte>;
for (i < v.length( { ... v[i]; } // iteration over length and explicit access still required for vectors
```

With generators, a sketch:
```
class Vector<T> {
  def for() -> T { // the "for" keyword designates a generator method
    for (i < length) yield this[i];
  }
}
var v: Vector<int>;
for (b in v) { ... } // a user value is iterable in a "for in" if it has a "for" method defined
```

## Function expressions, i.e. lambdas

I started sketching out a syntax for function expressions that can capture variables from the current environment; i.e. they'd be implemented as proper closures. After thinking about many alternatives, including just using the `=>` syntax like JavaScript, which I kind of like, I settled on using the `def` keyword to introducing a function expression. As for the name of the feature, I find that "function expression" is very self-explanatory, including to novice programmers, while "lambda" is understood by those who have a functional PL background. I prefer the self-explanatory, approachable term. But yeah, it's a lambda.

```
def m(f: int -> int) -> int { // a regular top-level method taking a function parameter
  return f(99);
}
// passing a function expression that adds 1 to its argument
m(def (x: int) -> int { return x + 1; });
// a much shorter version of the above, using => to skip return type, and inferred parameter type
m(def x => x + 1); 
```

## Double-arrow function bodies

Related to function expressions, but worthwhile in its own right, perhaps, is using `=>` to be able to define the body of method (and implicitly, its return type) as an expression. This helps with cleaning up additional line noise and making more one liners possible.

```
// before
class Foo {
  def m() -> int { return 33; }
  def n() -> int { return 44; }
  def s() -> int { return m() + n(); }
}
// after
class Foo {
  def m() => 33;
  def n() => 44;
  def s() => m() + n();
}
```

### Transparent structured data definitions

Ongoing work is adding a `struct` concept to Virgil. Structs are not a type in themselves; they are more like a view over memory, either a byte array (safe), or eventually, over raw, non-heap memory (unsafe). Structs are very explicit, requiring every field to have an offset and the overall struct size to be pre-declared.

```
struct S {
  +0 x: u32;
  +4 y: float;
  =12;
}
```

This will make interacting with software and hardware that has fixed memory layouts, like file formats, network packets, and external data structures, much, much easier.

**Comment by mosheduminer** [comment]
I'll think more about this and will probably have more feedback as I continue to use Virgil. But in the meantime I'll give some feedback here, based on my work on the Parser Combinator.

1. It would be convenient to have function expressions (which you already mentioned). In my work on the Parser Combinator, for every parser type, I needed to define an `impl` method with all required data as parameters, and then curry that the function to create the parser.

2. It would be convenient to be able to discard values when using `match`, right now every field in an ADT needs to be named. For example, instead of `Result(target, index, result) => ` it would be nice to be able to do `Result(target, _, _)`. I believe this is possible in python and rust (and probably other languages I'm not familar with).

3. Something that I thought about while implementing the Parser Combinator is the ability to define extensions on types (like [here](https://kotlinlang.org/docs/extensions.html)). My problem was that I wanted to define methods such as `.map`, `errorMap`, and `.chain` on the `ParserState -> ParserState` type, but this is of course impossible (you can't define methods on a function), so I needed to wrap the `fn: ParserState -> ParserState` into a class (I could have also made them standalone functions, but I preferred the DX of methods). There's not really a problem with this, it's just that I know classes are always allocated on the heap, and I would have preferred to avoid that. I'm not sure how generally useful this is though.

4. As I mentioned in the thread about JSON, compile time reflection (maybe also with decorators?) would be a really nice feature to have. Don't know what your plans about that are.


### Other (non-language) thoughts

1. Currently I'm using VS Code and it just treats virgil as text files. It would be really nice to have an LSP and VS Code extension, I saw this discussed in another thread, it's possible that I'd be able to work on this once we have JSON support. I think syntax highlighting, show definition on hover, jump to definition / usage, and autocomplete, would dramatically improve the coding experience.

2. I'm wondering what the approach would be to create bindings to something like sqlite? Or is there definitely no C compatibility?


My current goal with virgil is to get to the point where I can write a simple (single node) S3 compatible file server (mostly, I just want the most basic features). To that end I'm hoping to write a server application library (built on top of the networking work currently being done). It feels like something that is not very complex but will still be a milestone to get all the required pieces together.

More broadly I'm also curious about systems / embedded programming, but I wouldn't have any idea where to start learning about that. So, I'm focusing on higher level / web applications which I'm more familiar with.

**Reply by mosheduminer** [reply]
I know that none of these really improve virgil as a systems language. But honestly I have no idea about that, so I'm just mentioning some PL ideas that I had 😀.

**Reply by titzer** [reply]
Thanks for your thoughts.

Off the top of my head:

1. Parser work and the beginnings of semantic analysis for function expressions is done. Not sure I'll have much bandwidth for it, but the remaining bits are making the fat arrow return type inference (with cycle detection) work properly, not just for function expressions, but members, as well as doing proper scope analysis (upgrade MethodEnv). Then closure conversion to make it work propertly.

2. I've thought along similar lines and had the same syntax in mind. Are you up for jumping into the parser and making that work? It'd be a small adjustment to the syntax tree in `Vst.v3` as well. Overall, I think that'd be pretty easy.

3. It sounds like a useful feature for classes. Probably some syntactic wrinkles making work for other types.

4. A very large can of worms :-) What about source-code generation during compile-time and having a two-stage build? That wouldn't require any language changes.

Additionally,

1. VSCode integration would be great. I've met lots of people who would like to use it for Virgil. Also something I could use help with. I think the syntax highlighting part might be easy based off one of the grammars available for Virgil.

2. I think the "import" mechanism, which is currently used to define imports for Wasm modules, could be upgraded to describe interfaces in other languages, and combined with support for generating `.o` files, maybe Virgil could be linked into a larger C program.

**Reply by mosheduminer** [reply]
Discarding values - I can certainly try, though probably not for a least a few days. And I'll probably finish work on the JSON parser first.

Compile time macros / reflection / code generation is definitely a big can of worms. I only suggest it in terms of reflection because virgil already allows compile time code. It's definitely something that needs a lot of consideration and comparisons to other languages before embarking on it. Powerful features done wrong can probably ruin the language.

VSCode integration - I'll try to tackle this one after JSON and value discarding. I haven't made a VSCode extension before, but I'm hoping it won't be _too_ difficult 😂.

> I've met lots of people who would like to use it for Virgil.

I'm curious if you know how large the virgil userbase is (or people who would _like_ to use it)? And if you can disclose it 🙂.

**Reply by mosheduminer** [reply]
> What about source-code generation during compile-time and having a two-stage build

How would this allow for introspection, or does it not allow for it?

**Reply by k-sareen** [reply]
I've looked into LSP support ages ago and the sticking point was implementing the virgil source code parser all over again in your extension language. Since the grammar is always getting updated (the `/doc/virgil-grammar.ebnf` is already a bit stale I think), it would have been annoying to support. But now that Virgil has networking and soon will have JSON support, writing a native LSP _server_ in Virgil may not be hard? It's been a while since I looked into LSP stuff, but theoretically we can just use the actual Virgil src code parser instead of writing a new one (I personally really don't like writing parsers -- probably my least favourite compilation step 😆) and then use the JSON library to send data back-and-forth to the client (i.e. the VSCode extension).

**Reply by titzer** [reply]
Yes, I agree, using the official parser for the LSP would be best. I've tried to keep it fairly un-entangled (i.e. just creates Vst nodes), so we could further detangle it for reuse there.

This is actually very cool that it so happens that the two of you are working on the exact ingredients to make this happen :-)

**Reply by mosheduminer** [reply]
> This is actually very cool that it so happens that the two of you are working on the exact ingredients to make this happen :-)

Definitely 😃.

BTW for some reason I was under the impression that LSPs usually use stdin/stdout rather than sockets, although I believe either can be used?

**Reply by titzer** [reply]
> I'm curious if you know how large the virgil userbase is (or people who would _like_ to use it)? And if you can disclose it 🙂.

I can only go by the github stats, but there are probably a few hundred people who have tried it out and a few dozen people who have written more than a handful of lines of code. The heaviest users have been students working with me, like @k-sareen, who did a project with me after taking my compilers course at ANU. I have some new students spinning up next year, so investment in productivity and tooling will benefit them, too.

**Reply by k-sareen** [reply]
> BTW for some reason I was under the impression that LSPs usually use stdin/stdout rather than sockets, although I believe either can be used?

Seems like it's actually JSON RPC, so may still need a bit of work [1], but it should still just be using sockets. Though I don't remember if there was anything particular about the protocol used for communication (TCP vs UDP). Need to look at my old notes to check.

\[1]: https://microsoft.github.io/language-server-protocol/overviews/lsp/overview/

**Reply by mosheduminer** [reply]
Right, but the json-rpc 2.0 spec says that it's transport agnostic, so it technically doesn't require networking. It sounds like you've done more poking around than I have though. Anyway, it should be simple-ish to (de)serialize json-rpc once I finish a higher level of abstraction over the json parser.

**Reply by k-sareen** [reply]
True. But with a virgil LSP server and a Typescript LSP client, it may be easiest to just use a UDP connection. Best to survey what other languages do though. Should look at what smaller, niche languages like Zig do instead of some industrial behemoth like Java. I remember trying to read the Java/Rust/C LSP src code was a real pain. Maybe this has all improved in the last 1-2 years though.

**Reply by mosheduminer** [reply]
The zig LSP looks quite small, just a few files (ignoring the fact that the `lsp.zig` file is almost 10k lines 🙃). Anyway, they seem to be [using stdin/stdout](https://github.com/ziglibs/zig-lsp/blob/56caf2dd9ad218ea90f299f1575b7a240d2b4e14/src/main.zig#L39-L48).

**Reply by k-sareen** [reply]
That's cool! Thanks! Once JSON support has landed, we can look into LSP support.

**Reply by mosheduminer** [reply]
Oh, btw that 10k line file is generated by [a codegen](https://github.com/zigtools/zig-lsp-codegen).

**Reply by k-sareen** [reply]
Wew. For an MVP though I think only a few methods like `definitions` and `did{Open,Close}` are required.

**Comment by k-sareen** [comment]
This is a separate thing, but I feel like having an import statement + namespaces/modules will really help larger applications (for example `use lib.net.Net`) and streamline the build process.

Edit: As an aside, how does one get updates about the "Discussions" tab? Do you have to subscribe to something? I've never used it before.

**Reply by titzer** [reply]
There was some discussion in #79

**Reply by titzer** [reply]
Discussions is new to me; I am getting updates without having done anything special.

**Reply by mosheduminer** [reply]
You need to follow the repository if you want to get all updates. You can set a custom follow/notifications so you can choose to only get PRs/discussions/issues etc.

**Reply by k-sareen** [reply]
I apologize for being a Rust evangelist again, but I find its module system quite easy to do and you don't have to keep specifying what module etc. you are in every file. But it is very opinionated about directory structure which you may or may not want as part of the compiler. Rust does the `mod.rs` thing that you mentioned but you can also have self contained modules in a file. A bit like a component, really.

**Reply by mosheduminer** [reply]
BTW I found it quite unintuitive that while top level _functions_ are not exposed to other files, they need to be inside a components, _classes_ (and I suppose also other types / structures) can _only_ be top level (and are global).

**Reply by titzer** [reply]
There is some value in making some classes private to files, sure. I've often wanted to do that, e.g. for implementation details of things like hashtables. What new features would you find useful there?

**Reply by k-sareen** [reply]
Right. And on a similar note, it would be nice to have private types/classes as well. Since that could help in reducing the exposed surface for libraries/internals (for example the `IntPointer` class in the networking code). But then there is an added headache of resolving these private types. Need to keep track of modules at that point, I think?

**Reply by mosheduminer** [reply]
I think modules are a key part of this. It feels disturbing (and very C-like 🙂) for names to be global. But being able to declare some things as private would be nice.

**Reply by k-sareen** [reply]
Technically C has `static` for file-local names. But yes, I think private classes/types and even components maybe? are useful.

**Reply by mosheduminer** [reply]
Oh, and this is relevant to the LSP, because currently the compile resolves names purely based on whether they are available in the files listed on the command line. We can set up the LSP to follow the convention of checking for a `DEPS` file, but this is sorta related to the modules / namespacing question (though I think it doesn't actually matter either way).

**Reply by k-sareen** [reply]
Yeah the LSP integration with multi-file projects needs to be aware of the DEPS file convention. Though for MVP, maybe absolute barebones single-file project only? Though thinking about it, it should't be that hard to add DEPS support to the LSP.

**Reply by titzer** [reply]
Private classes/types would be easy to implement; the verifier already has a hierarchy of scopes for types (the `TypeEnv`), so introducing one for a file is not a big deal.

**Reply by mosheduminer** [reply]
Another, perhaps better, solution (vs private) would be to allow / require types to be part of components.

**Reply by titzer** [reply]
I whipped up private types here: https://github.com/titzer/virgil/commit/52681364ade309a1934b0d1b2a52e8dc4d035139

**Reply by k-sareen** [reply]
Oh wow! Very cool! I'll give it a whirl. Thanks!

**Reply by mosheduminer** [reply]
As far as I can tell, types can be made private, but can't be _namespaced_. Would it be possible to namespace a class to a component (given the lack of any other namespace mechanism)? Say something like this:

```
private class MyPrivateClass {}

component SomeComponent {
    def TheClass = MyPrivateClass;
}
```

Which currently doesn't work, because "type "MyPrivateClass" cannot be used as an expression".

**Reply by titzer** [reply]
Yeah, I'd like to introduce a notion of modules or namespaces at some point.

**Comment by k-sareen** [comment]
Also I think non-nullable types and some form of error propagation would be very useful as well, as we discussed in #111. Specially a native error type + some kind of operator for quickly propagating errors up the call stack, like the `?` operator.

**Comment by titzer** [comment]
I am thinking of a construct to make initialization of large, complex structures with named members easier. Some ideas:

# `with` expression

```
class Config {
  var name: string;
  var size: int;
  var doMore: bool;
  var extraThings: Array<int>;
}

def x = Config.new() with {
  name = "foo";
  size = 44;
  doMore = false;
  extraThings = [88, 99, 1000];
};
```

The idea of the `with` expression is that it takes an expression on the left and applies the updates to the fields given in the curlies `{ ... }`. For class types with mutable members, it takes an object, mutates the original object and returns it. For tuples, data types, or ADTs, it would create a new value with the updated fields.

It allows for having lots of parameters via a config object:

```
class Args { ... }
def m(x: Args) {
  ...
}
var r1 = m(Args.new() with { name = "string"; size = 66; });
// or, eliding the .new():
var r2 = m(Args with {name = "myname"; size = 777; });
```

Wdyt?

**Reply by k-sareen** [reply]
Interesting. Certainly improves readability for large constructors. An alternative could be to allow keyword arguments. Not arbitrary ones like in python, but something like:

```
class A {
  var a: int;
  var b: Array<int>;
  new(a, b) { }
}

def a = A.new(a=10, b=[11, 12, 13]);
```

**Reply by k-sareen** [reply]
Also: in the `with` proposal, do you have to provide all the fields or can some fields be omitted? As in, is this permissible in your example:

```
def x = Config.new() with {
  name = "foo";
  size = 44;
};
```

**Reply by titzer** [reply]
Yes, if you omit them, they get their default values (which could be from the field's initializer).

**Reply by k-sareen** [reply]
Oh that sounds pretty cool! Will a `class` variable end up as `null`? I guess so, right? since it's the default for an object. Might be ugly if you accidentally use it without initializing. Maybe only allow primitives/arrays/ADTs variables to be set in a `with` clause.

It might also be worth considering the use of some other keyword if you plan on adding scopes (/[structured concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/)) in the future as `with` is a common keyword to use in such a case.

**Reply by mosheduminer** [reply]
Hold up, are there already plans on structured concurrency? I actually made a (very) minimal library based on the rust async book.

**Reply by titzer** [reply]
I haven't made any concrete plans. I keep seeing that each kind of concurrency has its uses. `fork`/`join` parallelism is really useful in some contexts, while co-routines or generators are useful in others, and in others effect handlers are the thing. Still in other situations sometimes you just want a thread. How to reconcile all of these possible directions? Unclear. What I usually do is something pragmatic  and low-level that lets me experiment with the next level to see how that should be blessed and integrated into the language.

**Reply by k-sareen** [reply]
Arguably a systems language should provide a 1:1 threading model (i.e. 1 language thread = 1 OS thread), with either async primitives or some other coroutine facility to make an M:N threading model if possible. Or this is at least from my experience with a bunch of different systems-y languages. M:N threading models are great to think in though. And structured concurrency lends itself really well to M:N threading models. Just my thoughts.
